#12277 Fix usage of hash argument to attr.s

iwane-pl opened the pull request:
## Scope and purpose

Fixes #12277 

According to [attr.s documentation](https://www.attrs.org/en/stable/hashing.html), `hash` shouldn't be used directly. Either `frozen` or `eq` should be manipulated, so used `frozen` wherever the code called for `hash=True` and `eq=False` (which was
present already) wherever it called for `hash=False`.

iwane-pl committed:
Fix usage of hash argument to attr.s

According to attr.s documentation, `hash` shouldn't
be used directly. Either `frozen` or `eq` should
be manipulated, so used `frozen` wherever the code
called for `hash=True` and `eq=False` (which was
present already) wherever it called for `hash=False`.

adiroiban commented on src/twisted/web/newsfragments/12277.bugfix:
```diff
@@ -0,0 +1 @@
+twisted.internet.address no longer raises DeprecationWarning when used with attrs>=24.1.0. As a side effect, `IPV4Address`/`IPV6Address`/`HostnameAddress` instances are now `frozen` (i.e. immutable).
```
I agree that the addresses should be immutable ... but this looks like  a backard incompatible change.

It looks like there is also a performance penalty for enforcing immutable instances.

I think that for this PR, is best to just use `unsafe_hash`.
My understanding is that `unsafe_hash` is just the rename of the old `hash`

I think that `unsafe_hash` was added in `attrs 22.2.0`

Right now we support `attrs >= 21.3.0`

https://github.com/twisted/twisted/blob/bb5ac14f8941ec7d4823e402d4d7d47c663ecc16/pyproject.toml#L39

iwane-pl replied:
Fair point. I wasn't sure which way is the intended way :) Certainly I don't want to dive in the performance, as I already have too many contexts to keep in the head.

Back to `unsafe_hash` it is, then.

adiroiban reviewed (dismissed):
Thanks for the changes.

My suggestion is to continue to use `unsafe_hash` and update the dependency for Twisted to `attrs >= 22.2.0`


-------

If we want to make the addresses immutable, that should be a separate issue and we should discuss the performance implication.

iwane-pl committed:
Replace `hash` with `unsafe_hash`

As discussed in the #12278 review, this is the preferred
option now, as introducing immutability to *Address
classes causes a performance hit.

twm reviewed (approved):
Awesome, I was just hitting this and about to fix it myself. :smile: 

Thanks @iwane-pl!

adiroiban commented on pyproject.toml:
```diff
@@ -36,7 +36,7 @@ dependencies = [
     "Automat >= 0.8.0",
     "hyperlink >= 17.1.1",
-    "attrs >= 21.3.0",
+    "attrs >= 22.2.0",
```
When we change the dependencies here, we should also update the minimum dependency test environment in tox

https://github.com/twisted/twisted/blob/f59bb561e0393458e8c113a822f8cff8a3ed69f7/tox.ini#L80

------

I don't know why the tests were not failing on mindeps... I will investigate

-------
Job output ... `attrs == 21.3.0` is requested... but the env ends up with a newer version

this needs investigation in a separate ticket

https://github.com/twisted/twisted/actions/runs/10254839493/job/28386706743?pr=12278#step:10:40

```
mindeps-withcov-posix installdeps: zope.interface == 5.0.0, constantly == 15.1, incremental == 24.7.0, Automat == 0.8.0, hyperlink == 17.1.1, attrs == 21.3.0, typing_extensions == 4.2.0, pyopenssl == 21.0.0, service_identity == 18.1.0, idna == 2.4, cryptography == 3.3, appdirs == 1.4.0, bcrypt == 3.1.3, h2 == 3.0, priority == 1.1.0, cython-test-exception-raiser == 1.0.2, PyHamcrest == 2.0, hypothesis == 6.56
mindeps-withcov-posix inst: /home/runner/work/twisted/twisted/.tox/.tmp/package/1/twisted-24.3.0.post0-py3-none-any.whl
mindeps-withcov-posix installed: alabaster==0.7.13,appdirs==1.4.0,astor==0.8.1,astroid==3.2.4,attrs==24.1.0 ....
```

adiroiban replied:
I have created a separate issue to fix this https://github.com/twisted/twisted/issues/12282

adiroiban commented:
Thanks Piotr for the fix and thanks Tom for the merge.
Happy for this cleanup.
